logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/slow/ExecuteSlowSyncForTooLongOfflineUseCase.kt

```diff
@@ -17,26 +17,51 @@
  */
 package com.wire.kalium.logic.sync.slow
 
+import co.touchlab.stately.concurrency.AtomicBoolean
+import com.wire.kalium.common.error.CoreFailure
+import com.wire.kalium.common.functional.Either
+import com.wire.kalium.logger.KaliumLogger
 import com.wire.kalium.logic.data.sync.SlowSyncRepository
 import io.mockative.Mockable
-import kotlinx.coroutines.flow.Flow
-import kotlinx.datetime.Instant
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.take
+import kotlin.time.Duration.Companion.seconds
 
 /**
  * Execute slow sync because the client was offline for too long.
- * It will wait until [slowSyncRequester] finishes or fails.
+ * It will wait until a slow sync is completed and return the instant of completion.
  */
 @Mockable
 internal interface ExecuteSlowSyncForTooLongOfflineUseCase {
-    suspend operator fun invoke(): Flow<Instant?>
+    suspend operator fun invoke(onComplete: suspend () -> Either<CoreFailure, Unit>)
 }
 
-internal class ExecuteSlowSyncForTooLongOfflineUseCaseImpl(private val slowSyncRepository: SlowSyncRepository) :
-    ExecuteSlowSyncForTooLongOfflineUseCase {
+internal class ExecuteSlowSyncForTooLongOfflineUseCaseImpl(
+    private val slowSyncRepository: SlowSyncRepository,
+    logger: KaliumLogger
+) : ExecuteSlowSyncForTooLongOfflineUseCase {
 
-    override suspend operator fun invoke(): Flow<Instant?> {
-        slowSyncRepository.setNeedsToPersistHistoryLostMessage(true)
-        slowSyncRepository.clearLastSlowSyncCompletionInstant()
-        return slowSyncRepository.observeLastSlowSyncCompletionInstant()
+    val logger = logger.withTextTag("ExecuteSlowSyncForTooLongOfflineUseCase")
+    val isRunning = AtomicBoolean(false)
+
+    override suspend operator fun invoke(onComplete: suspend () -> Either<CoreFailure, Unit>) {
+        if (!isRunning.value) {
```

Comment:
We need to handle state locally, it can be the case as in my tests, that the ws connection drops and the receives again the missed notification, since the slow sync is running, we need to not retrigger it again.